jvm-packages/xgboost4j-spark-gpu/src/main/scala/ml/dmlc/xgboost4j/scala/rapids/spark/GpuPreXGBoost.scala

```diff
@@ -0,0 +1,572 @@
+                  val dm = new DMatrix(cudfColumnBatch, missing, 1)
+                  if (dm == null) {
+                    Iterator.empty
+                  } else {
+                    try {
+                      currentBatch = new ColumnarBatch(
+                        GpuColumnVector.extractColumns(table, dataTypes).map(_.copyToHost()),
+                        table.getRowCount().toInt)
+                      val rowIterator = currentBatch.rowIterator().asScala
+                        .map(toUnsafe)
+                        .map(converter(_))
+                      predictFunc(bBooster, dm, rowIterator)
+
+                    } finally {
+                      dm.delete()
+                    }
+                  }
+                } finally {
+                  feaTable.close()
+                }
+              }
+            } else {
+              iter = null
+            }
+          }
+
+          override def hasNext: Boolean = {
+            val itHasNext = iter != null && iter.hasNext
+            if (!itHasNext) { // Don't have extra Row for current ColumnarBatch
+              loadNextBatch()
+              iter != null && iter.hasNext
+            } else {
+              itHasNext
+            }
+          }
+
+          override def next(): Row = {
+            if (iter == null || !iter.hasNext) {
+              loadNextBatch()
+            }
+            if (iter == null) {
+              throw new NoSuchElementException()
+            }
+            iter.next()
+          }
+        }
+    }
+
+    bOrigSchema.unpersist(blocking = false)
+    bRowSchema.unpersist(blocking = false)
+    bBooster.unpersist(blocking = false)
+    dataset.sparkSession.createDataFrame(rowRDD, schema)
+  }
+
+  /**
+   * Transform schema
+   *
+   * @param est supporting XGBoostClassifier/XGBoostClassificationModel and
+   *                 XGBoostRegressor/XGBoostRegressionModel
+   * @param schema   the input schema
+   * @return the transformed schema
+   */
+  override def transformSchema(
+      est: XGBoostEstimatorCommon,
+      schema: StructType): StructType = {
+
+    val fit = est match {
+      case _: XGBoostClassifier | _: XGBoostRegressor => true
+      case _ => false
+    }
+
+    val Seq(label, weight, margin) = GpuUtils.getColumnNames(est)(est.labelCol, est.weightCol,
+      est.baseMarginCol)
+
+    GpuUtils.validateSchema(schema, est.getFeaturesCols, label, weight, margin, fit)
+  }
+
+  /**
+   * Repartition all the Columnar Dataset (training and evaluation) to nWorkers,
+   * and assemble them into a map
+   */
+  private def prepareInputData(
+      trainingData: ColumnDataBatch,
+      evalSetsMap: Map[String, ColumnDataBatch],
+      nWorkers: Int,
+      isCacheData: Boolean): Map[String, ColumnDataBatch] = {
+    // Cache is not supported
+    if (isCacheData) {
+      logger.warn("Dataset cache is not support for GPU pipeline!")
```

Comment:
What is a cache in the context of the spark package?

Reply:
@wbo4958

Reply:
cache which means cache the spark computation result in local storage.  Gpu Pipeline only accelerate the Dataset cache instead of RDD cache. so for now we just disable the cache.

will figure out a way to handle this

Reply:
Got it.  That sounds fine.